### PR TITLE
sasarkar/Map list to tensors

### PR DESCRIPTION
**Pull request by ssarkar2**
# What does this PR do?
**Original issue**
resuming for multi card fails

```python ../gaudi_spawn.py --world_size 2 --use_deepspeed run_clm.py --run_name initial_run --model_name_or_path openai-community/gpt2 --dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 --block_size 4096 --per_device_train_batch_size 8 --gradient_accumulation_steps 8 --per_device_eval_batch_size 4 --do_train --do_eval --learning_rate 4e-4 --output_dir resume_exp3_2x --gaudi_config_name Habana/gpt2 --use_habana --use_lazy_mode --use_hpu_graphs_for_inference --gradient_checkpointing --use_cache False --throughput_warmup_steps 3 --deepspeed ds_zero2.json --overwrite_output_dir --num_train_epochs 1 --logging_dir resume_exp3_2x --logging_steps 1 --streaming --max_steps 7 --save_strategy steps --save_steps 5```

```python ../gaudi_spawn.py --world_size 2 --use_deepspeed run_clm.py --run_name initial_run --model_name_or_path openai-community/gpt2 --dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 --block_size 4096 --per_device_train_batch_size 8 --gradient_accumulation_steps 8 --per_device_eval_batch_size 4 --do_train --do_eval --learning_rate 4e-4 --output_dir resume_exp3_2x --gaudi_config_name Habana/gpt2 --use_habana --use_lazy_mode --use_hpu_graphs_for_inference --gradient_checkpointing --use_cache False --throughput_warmup_steps 3 --deepspeed ds_zero2.json --num_train_epochs 1 --logging_dir resume_exp3_2x --logging_steps 1 --streaming --max_steps 60 --save_strategy steps --save_steps 5 --resume_from_checkpoint resume_exp3_2x/checkpoint-5```

The second command fails with error: `RuntimeError: Error when trying to cast Long to Int, Input values range [-5337826792292336652, 7161124111582705975] exceeds Int range [-2147483648, 2147483647]`

**Reason**

When we resume, [this](https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/trainer.py#L883) code is run: `skip_first_batches`

This calls this [line](https://github.com/huggingface/accelerate/blob/cd7df4117d92f660965d5f737364395b5693a535/src/accelerate/data_loader.py#L1085).

When iterating over the dataloader, we hit this [line](https://github.com/huggingface/accelerate/blob/cd7df4117d92f660965d5f737364395b5693a535/src/accelerate/data_loader.py#L672).

This call finally initializes tensor [here](https://github.com/huggingface/accelerate/blob/cd7df4117d92f660965d5f737364395b5693a535/src/accelerate/utils/operations.py#L240)

This creates a `torch.empty` tensor, which is uninitialized, and with dtype=int64

HPU doesn't support int64, and the uninitialized int64 tensor has large out of bounds number, which cannot be converted to int32, hence the error.

**Solution**
The training dataset is a list of dictionaries, whose values are lists of integers. These lists are converted into int64 inside `Trainer`. To prevent that, I map the dataset into int32 tensors.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: examples/language-modeling/run_clm.py [hunk 1]
```diff
@@ -592,9 +592,15 @@ def group_texts(examples):
...
 
     if training_args.do_train:
+
+        def tensor_mapper(x):
+            return {i: torch.tensor(x[i], dtype=torch.int32) for i in x}
+
         if "train" not in tokenized_datasets:
             raise ValueError("--do_train requires a train dataset")
         train_dataset = lm_datasets["train"]
+        if training_args.resume_from_checkpoint != '':
```
Comment (regisss) [inline examples/language-modeling/run_clm.py:602]: `training_args.resume_from_checkpoint` may be None, as it's the default value if it is not specified. So I think we should do:
```suggestion
        if training_args.resume_from_checkpoint is not None and training_args.resume_from_checkpoint != '':
```

**Review by regisss — COMMENTED** [review]
Good catch!

File: examples/language-modeling/run_clm.py [hunk 2]
```diff
@@ -592,9 +592,14 @@ def group_texts(examples):
...
 
     if training_args.do_train:
+
+        def tensor_mapper(x):
+            return {i: torch.tensor(x[i], dtype=torch.int32) for i in x}
+
         if "train" not in tokenized_datasets:
             raise ValueError("--do_train requires a train dataset")
         train_dataset = lm_datasets["train"]
+        train_dataset = train_dataset.map(tensor_mapper)
```
[the remark this answers is not in the document]
Comment (regisss) [reply examples/language-modeling/run_clm.py]: Yes I agree, only if the dataset is an iterable

[the remark this answers is not in the document]
Comment (ssarkar2) [reply examples/language-modeling/run_clm.py]: @regisss :
Actually I forget what I meant by this, and what you mean in your reply. could you please clarify if u suggest some change here?

[the remark this answers is not in the document]
Comment (regisss) [reply examples/language-modeling/run_clm.py]: I was thinking that, in some cases, the dataset may not be an iterable (and so `.map` would not work) but actually I'm not even sure about that. Let's merge it that way and we'll see if we encounter any issue in the future.

**Pull request merged**